portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,35 +24,81 @@
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comment data. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private ArrayList<String> comments;
-
-  @Override
-  public void init() {
-    comments = new ArrayList<String>();
-  }
-
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String userComment = request.getParameter("text-input");
-    if (userComment == null) {
-      userComment = "";
+    if (userComment != null) {
+      storeComment(userComment);
     }
-    comments.add(userComment);
     response.sendRedirect("/index.html");
   }
 
+  /**
+    * Stores user comment as entity in datastore.
+    */
+  private void storeComment(String userComment) {
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("text", userComment);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PreparedQuery results = getAllComments();
+    
+    ArrayList<String> comments = new ArrayList<String>();
+    int numberToDisplay = getNumberToDisplay(request); 
+    int numberDisplayed = 0;
+
+    for(Entity comment : results.asIterable()) {
+      String commentText = (String) comment.getProperty("text");
+      comments.add(commentText);
+      numberDisplayed += 1;
+      if (numberDisplayed == numberToDisplay) {
+        break;
+      }
+    }
+
     String json = convertToJson(comments); 
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
+  /** 
+    * Returns number of comments to display, based on user's 
+    * selection. 
+    */
+  private int getNumberToDisplay(HttpServletRequest request) {
+    String parameterString = (String) request.getParameter("numberToDisplay");
+    int numberToDisplay;
+    try {
+      numberToDisplay = Integer.parseInt(parameterString);
+    } catch (Exception e) {
+      System.err.println("Could not convert to int: " + parameterString +
+          "\nSetting value to default: 5");
```

Comment:
Instead of always defaulting to 5, there may be different cases where you do or don't want to do so. 

Ex1: User omits "numberToDisplay" parameter. This is a good time to default.
Ex2: User sets the param to a 64bit number (would also through a parse error, even though unlikely a user would actually do this). This would likely be OK throwing an Invalid Argument Exception
Ex3: User sets the param to a letter. Would also probably be better throwing an Invalid Argument Exception.

Reply:
I'm realizing that this part of the code I wrote is a bit confusing, sorry! In this case, the parameter returned by getParameter("numberToDisplay") isn't something the user typed in; I actually have the user selecting a value from a dropdown menu, so right now the only values that could be passed to this function are 5, 10, 15, and 20. Assuming I've done everything right on my end, this parseInt() should never fail. With this in mind, is it confusing that I even try to catch this error? Should I remove the try/catch block and simply let the program crash if parseInt() fails?

Reply:
I would consider the Preconditions library for this. This gives the client application that the information was an invalid argument, rather than an internal server error
  
and using checkArgument(). This would be a standard practice within a normal application, but not technically required for this project. 


https://guava.dev/releases/19.0/api/docs/com/google/common/base/Preconditions.html

Reply:
It looks like checkArgument() accepts a boolean expression as input. I'm having trouble finding an easy way to check that a string can be converted to an integer without using parseInt + a try/catch block. Since it looks like checkArgument() simply throw an IllegalArgumentException when the condition is not met, should I modify my code so that my catch {} block throw an IllegalArgumentException?